cirq/testing/order_tester.py

```diff
@@ -0,0 +1,230 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""A utility class for testing ordering methods.
+
+To test an ordering method, create an OrderTester and add several
+equivalence groups or items to it. The order tester will check that
+the items within each group are all equal to each other, and every new
+added item or group is strictly ascending with regard to the previously
+added items or groups.
+
+It will also check that a==b implies hash(a)==hash(b).
+"""
+
+from typing import Any
+
+from cirq.testing.equals_tester import EqualsTester
+
+
+class OrderTester(EqualsTester):
+    """Tests ordering against user-provided disjoint ordered groups or items."""
+
+    def __init__(self):
+        self.groups = [(ClassSmallerThanEverythingElse(),)]
+
+    def _verify_ascending(self, v1, v2):
+        """Verifies that (v1, v2) is a strictly ascending sequence."""
+        lt_1 = NotImplemented
```

Comment:
I *think* this method can just be the eight asserts I listed earlier (four comparisons * two outcomes). The correct `NotImplemented` behavior will be checked by a new helper method `_assert_not_implemented_vs_unknown` that is only called when that should be relevant.